lms/djangoapps/grades/models.py

```diff
@@ -311,74 +268,95 @@ def read_grade(cls, user_id, usage_key):
+            course_key: The course identifier for the desired grades
+        """
+        return cls.objects.select_related('visible_blocks').filter(
             user_id=user_id,
-            usage_key=usage_key,
+            course_id=course_key,
         )
 
-        grade.update(
-            course_version=course_version,
-            subtree_edited_timestamp=subtree_edited_timestamp,
-            earned_all=earned_all,
-            possible_all=possible_all,
-            earned_graded=earned_graded,
-            possible_graded=possible_graded,
-            visible_blocks=visible_blocks,
+    @classmethod
+    def update_or_create_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of those methods.
+        """
+        cls._prepare_params_and_visible_blocks(kwargs)
+
+        user_id = kwargs.pop('user_id')
+        usage_key = kwargs.pop('usage_key')
+
+        grade, _ = cls.objects.update_or_create(
+            user_id=user_id,
+            course_id=usage_key.course_key,
+            usage_key=usage_key,
+            defaults=kwargs,
         )
+        return grade
+
+    @classmethod
+    def create_grade(cls, **kwargs):
+        """
+        Wrapper for objects.create.
+        """
+        cls._prepare_params_and_visible_blocks(kwargs)
+        return cls.objects.create(**kwargs)
+
+    @classmethod
+    def bulk_create_grades(cls, grade_params_iter, course_key):
+        """
+        Bulk creation of grades.
+        """
+        if not grade_params_iter:
+            return
+
+        map(cls._prepare_params, grade_params_iter)
+        VisibleBlocks.bulk_get_or_create([params['visible_blocks'] for params in grade_params_iter], course_key)
+        map(cls._prepare_params_visible_blocks_id, grade_params_iter)
 
-    def update(
-            self,
-            course_version,
-            subtree_edited_timestamp,
-            earned_all,
-            possible_all,
-            earned_graded,
-            possible_graded,
-            visible_blocks,
-    ):
-        """
-        Modify an existing PersistentSubsectionGrade object, saving the new
-        version.
-        """
-        visible_blocks_hash = VisibleBlocks.objects.hash_from_blockrecords(BlockRecordList.from_list(visible_blocks))
-
-        self.course_version = course_version or ""
-        self.subtree_edited_timestamp = subtree_edited_timestamp
-        self.earned_all = earned_all
-        self.possible_all = possible_all
-        self.earned_graded = earned_graded
-        self.possible_graded = possible_graded
-        self.visible_blocks_id = visible_blocks_hash  # pylint: disable=attribute-defined-outside-init
-        self.save()
-        log.info(u"Persistent Grades: Grade model updated: {0}".format(self))
+        return cls.objects.bulk_create([PersistentSubsectionGrade(**params) for params in grade_params_iter])
```

Comment:
I had initially, but was running into issues when re-using the generator multiple times.  So I'm sorting to a list for now.  We do not expect the number of Subsections to exceed much more than 25 or so.